### ISSUE TITLE
TextFormat: Return value of TextFormat::colorize() must be of the type string

### ISSUE THREAD

**Issue by zKoz210** [pmmp/PocketMine-MP#2130]
### Issue description
Use emoticons causing the error in the replace color

### Steps to reproduce the issue

1. Write any message and add emoji (Works only on Android)

### OS and versions
* PocketMine-MP: https://github.com/pmmp/PocketMine-MP/commit/1fc388d6debf0d38e529d5161a4d0f28fa86d075
* PHP: 7.2.3
* Server OS: CentOS 7.2.3
* Game version: PE/Win10

### Crashdump, backtrace or other files
Video: [20180402_202814.zip](https://github.com/pmmp/PocketMine-MP/files/1869367/20180402_202814.zip)

```
Невозможно передать событие 'pocketmine\event\player\PlayerChatEvent' в 'Chat v1.0.0': Return value of pocketmine\utils\TextFormat::colorize() must be of the type string, null returned на \Chat\CListener
TypeError: "Return value of pocketmine\utils\TextFormat::colorize() must be of the type string, null returned" (EXCEPTION) in "src/pocketmine/utils/TextFormat" at line 92
```

**Comment by dktapps** [comment]
My soft keyboard doesn't permit typing emojis into the chat, so I can't reproduce this issue.

**Comment by dktapps** [comment]
I am also not able to reproduce this using alternative methods, so I'm going to close this.

**Comment by lukeeey** [comment]
How can you not type emojis? I can using the default Android 5.1 keyboard

**Comment by zKoz210** [comment]
@dktapps Please reopen, I had some problems and I couldn't keep track of your project. This bug is still reproduce on Gboard keyboard

An example of how to debug:
```
public function onPlayerChatEvent(PlayerChatEvent $event){
	var_dump($event->getMessage());
	var_dump(bin2hex($event->getMessage()));
}

```
**Debug:**
![image](https://user-images.githubusercontent.com/28255085/41460939-e31e40b8-7096-11e8-881f-b28baa8a4ee6.png)

```
<?php

$hex2bin = hex2bin("eda0bdedb894");

echo colorize($hex2bin);

function colorize(string $string, string $placeholder = "&") : string{
	return preg_replace('/' . preg_quote($placeholder, "/") . '([0-9a-fk-or])/u', '§' . '$1', $string);
}
```

Result:
```
FATAL ERROR Uncaught TypeError: Return value of colorize() must be of the type string, null returned in /home4/phptest/public_html/code.php70(5) : eval()'d code:8 Stack trace: #0 /home4/phptest/public_html/code.php70(5) : eval()'d code(5): colorize('\xED\xA0\xBD\xED\xB8\x94') #1 /home4/phptest/public_html/code.php70(5): eval() #2 {main} thrown on line number 8
```

**Comment by dktapps** [comment]
This appears to be a UTF issue - when removing the `u` modifier, it works correctly. https://stackoverflow.com/questions/1435190/preg-replace-null-result

**Comment by zKoz210** [comment]
This will be resolved?

**Comment by dktapps** [comment]
Nope

**Comment by dktapps** [comment]
There are two solutions I can think of for this bug:

- Add an extra `$utf` parameter
- Have `colorize()` guess and try to match again if `preg_replace()` returned null. There are several other `TextFormat` functions which probably ought to follow this behaviour too - try UTF first, then fallback to ASCII if it fails. I'm not sure if this might produce unexpected behaviour.

This would not account for things like widestrings though.

**Comment by dktapps** [comment]
Further investigation reveals this is because the string that's being passed in is invalid UTF-8.

A few questions:
- Is this still reproducible with latest versions?
- If it is, can you capture a hex-encoded buffer of `TextPacket` as it arrives, and paste it here? I suspect the bug is coming from `TextFormat::clean()`.

**Comment by SOF3** [comment]
I doubt on the need of the `u` flag. According to the UTF-8 standard, all continuation bytes are U+10xxxxxx, while all leading bytes for multibyte characters start with U+11xxxxxx, and all single byte characters start with U+0xxxxxxx. If we assume the content is in UTF-8, and the needle and haystack strings are valid UTF-8, there is no chance that Unicode-unsafe replacement would occur even by naive bytewise replacement. But if the strings are invalid UTF-8, then GIGO and it's not our responsibility to fix.
In short, just remove the `u` character and end of story.

**Comment by dktapps** [comment]
I'm assuming this was fixed by https://github.com/pmmp/PocketMine-MP/commit/c433fad0a7077c8402eed5a68f037ea110e20446 , judging by the lack of feedback.
